include/ringmesh/basic/attributes.h

```diff
@@ -0,0 +1,1024 @@
+                }
+            }
+            ringmesh_assert( store_->get_store().elements_type_matches(
+                typeid( T ).name() ) );
+        }
+
+        /**
+         * \brief Destroys this attribute in the AttributesManager.
+         * \details On exit, the attribute is no-longer accessible in
+         *  the AttributesManager, its name is available again, and
+         *  this attribute is in the unbound state.
+         */
+        void destroy()
+        {
+            manager_.delete_attribute_store( store_ );
+        }
+
+        void set_constant_value( T value )
+        {
+            store_->set_store( new ConstantStore< T > );
+            set_value( 0, value );
+        }
+
+        /**
+         * \brief Tests whether an attribute with the specified name and with
+         *  corresponding type exists in an AttributesManager.
+         * \param[in] manager a reference to the AttributesManager
+         * \param[in] name the name of the attribute
+         */
+        static bool is_defined(
+            AttributesManager& manager, const std::string& name )
+        {
+            auto store = manager.find_attribute_store( name );
+            return ( store != nullptr
+                     && store->elements_type_matches( typeid( T ).name() ) );
+        }
+
+        /**
+         * \brief Gets the size.
+         * \return The number of items in this attribute.
+         */
+        index_t size() const
+        {
+            return store_->get_store().size();
+        }
+
+        /**
+         * \brief Gets a modifiable element by index
+         * \param [in] i index of the element
+         * \param [in] value to set at the \p i%th element
+         */
+        void set_value( index_t i, const T& value )
+        {
+            ringmesh_assert( i < size() );
+            *reinterpret_cast< T* >( store_->element( i ) ) = value;
+        }
+
+        /**
+         * \brief Gets an element by index
+         * \param [in] i index of the element
+         * \return a const reference to the \p i%th element
+         */
+        const T& operator[]( index_t i ) const
+        {
+            return value( i );
+        }
+
+        /**
+        * \brief Gets an element by index
+        * \param [in] i index of the element
+        * \return a const reference to the \p i%th element
+        */
+        const T& value( index_t i ) const
+        {
+            ringmesh_assert( i < size() );
+            return *reinterpret_cast< const T* >( store_->element( i ) );
+        }
+
+        /**
+         * \brief Sets all the elements in this attribute
+         *   to a specified value.
+         */
+        void fill( const T& value )
+        {
+            for( auto i : range( store_->get_store().nb_elements() ) )
+            {
+                set_value( i, value );
+            }
+        }
+
+    protected:
+        AttributesManager& manager_;
+        AttributeStore* store_;

```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Make this member variable "private". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=cpp%3AS3656)